File: Graphs/WeightedDirected/src/weighteddirected/Queue.java
package weighteddirected;

class QNode {

    public String data;
    public QNode next;

    public QNode(String data) {
        this.data = data;
        this.next = null;
    }
}

public class Queue {

    private QNode head;

    public Queue() {
        this.head = null;
    }

    public boolean isEmpty() {
        return this.head == null;
    }

    public void enqueue(String key) {
        if (this.head == null) {
            this.head = new QNode(key);
            return;
        }

        QNode node = this.head;
        while (node.next != null) {
            node = node.next;
        }

        node.next = new QNode(key);
    }

    public QNode dequeue() {
        if (this.head == null) {
            return null;
        }
        
        QNode node = this.head;
        this.head = this.head.next;
        return node;
    }
}

File: StackQueue/C++/Queue/Queue/Queue.cpp
#include <iostream>

class Queue {
	private:
		int front;
		int rear;
		int size;
		int* arr;

	public:
		Queue(int size) {
			this->front = -1;
			this->rear = -1;
			this->size = size;
			this->arr = new int[size];
			
			for (int i = 0; i < size; i++) {
				this->arr[i] = 0;
			}
		}

		bool isEmpty() {
			return this->front == -1 && this->rear == -1;
		}

		bool isFull() {
			return this->rear == this->size - 1;
		}

		bool enqueue(int value) {
			if (this->isFull())
				return false;
			
			if (this->isEmpty()) {
				this->rear = this->front = 0;
				this->arr[this->rear] = value;
			}
			else {
				this->rear++;
				this->arr[this->rear] = value;
			}
			return true;
		}

		int dequeue() {
			if (this->isEmpty()) {
				return 0;
			}

			if (this->front == this->rear) {
				int x = this->arr[this->front];
				this->rear = this->front = -1;
				this->arr[this->front] = 0;
				return x;
			}
			else {
				int x = this->arr[this->front];
				this->arr[this->front] = 0;
				this->front++;
				return x;
			}
		}

		int count() {
			return (this->rear == -1 && this->front == -1) ? 0 : this->rear - this->front + 1;
		}
};